### PR TITLE
simple command costmap api - update few functions

### PR DESCRIPTION
**Pull request by JacksonK9**
Updated-by: Jaehun Kim <k9632441@gmail.com>

* finish task B and tested on gazebo

* Update nav2_simple_commander/nav2_simple_commander/costmap_2d.py

---

## Basic Info

| Info | Please fill out this column |
| ------ | ----------- |
| Ticket(s) this addresses   | #3151  |
| Primary OS tested on | Ubuntu |
| Robotic platform tested on | gazebo |

---

## Description of contribution in a few bullet points
* Update costmap API for OccupancyGrid messages coming into the Simple Commander API
* Implementation of the following few functions
    * getCostIdx
    * getCostXY
    * setCost
    * mapToWorld
    * worldToMap
    * getIndex

## Description of documentation updates required from your changes
* Added new api, documentation on this api might be required in the simple commander page documentation.
---

## Future work that may be required in bullet points
* Implement and test a line iterator to march through the costmap given 2 points ancd check for collision
* Use that line iterator + costmap object + footprint input to check for collisions

## Related issue
* #3151 
* #3159 

#### For Maintainers: 
- [ ] Check that any new parameters added are updated in navigation.ros.org
- [ ] Check that any significant change is added to the migration guide
- [ ] Check that any new features **OR** changes to existing behaviors are reflected in the tuning guide
- [ ] Check that any new functions have Doxygen added
- [ ] Check that any new features have test coverage
- [ ] Check that any new plugins is added to the plugins page
- [ ] If BT Node, Additionally: add to BT's XML index of nodes for groot, BT package's readme table, and BT library lists

### CODE REVIEW

File: nav2_simple_commander/nav2_simple_commander/costmap_2d.py [hunk 1]
```diff
@@ -71,3 +71,72 @@ def getGlobalFrameID(self):
     def getCostmapTimestamp(self):
         """Get costmap timestamp."""
         return self.costmap_timestamp
+
+    def getCost(self, _index: int = None, _mx: int = None, _my: int = None) -> np.int8:
```
Comment (SteveMacenski) [inline nav2_simple_commander/nav2_simple_commander/costmap_2d.py:75]: Why are these starting with underscores? For all functions, I don't think that's proper

Comment (SteveMacenski) [inline nav2_simple_commander/nav2_simple_commander/costmap_2d.py:75]: Create 2 separate implementations for index vs mx,my to have API parity with C++ usage. No reason to confuse users having to `null` out `index` to use the mx/my API

**Review by SteveMacenski — CHANGES_REQUESTED** [review]
Generally looks good - small changes requested.

Make sure to update the table in https://navigation.ros.org/commander_api/index.html too!

File: nav2_simple_commander/nav2_simple_commander/costmap_2d.py [hunk 2]
```diff
@@ -1,6 +1,7 @@
 #! /usr/bin/env python3
 # Copyright 2021 Samsung Research America
 # Copyright 2022 Stevedan Ogochukwu Omodolor
+# Copyright 2022 Jaehun Jackson Kim
```
Comment (SteveMacenski) [inline nav2_simple_commander/nav2_simple_commander/costmap_2d.py:4]: If you're doing this on behalf of a company, you should list your company instead, just FYI :smile:

Comment (JacksonK9) [reply nav2_simple_commander/nav2_simple_commander/costmap_2d.py:4]: Okay, Thank you for your information, It is my first time, So I missed it. I will change 😊

File: nav2_simple_commander/nav2_simple_commander/costmap_2d.py (hunk 1 above)
Comment (JacksonK9) [reply nav2_simple_commander/nav2_simple_commander/costmap_2d.py:75]: Okay, I will remove all the underscore

Comment (JacksonK9) [reply nav2_simple_commander/nav2_simple_commander/costmap_2d.py:75]: I understand your comment as "split as two seperate functions like getCostFromIndex and getCostFromCell" 
Is it right?
I want to adopt the method overloading, but python does not support naturally.
If I use the multipledispatch package, I can use method overloading. 
I checked this post 
https://www.geeksforgeeks.org/python-method-overloading/amp/
If you have another idea, please let me know little more specific.
Thank you for your comment 👍

Comment (SteveMacenski) [reply nav2_simple_commander/nav2_simple_commander/costmap_2d.py:75]: Ah, I did not konw that. How about `getCostIdx` and `getCostXY`?

Comment (JacksonK9) [reply nav2_simple_commander/nav2_simple_commander/costmap_2d.py:75]: Okay that is better and simpler.
I will change it.
Thank you for your support 😆

Comment (JacksonK9) [reply nav2_simple_commander/nav2_simple_commander/costmap_2d.py]: remove all the underscores

Comment (JacksonK9) [reply nav2_simple_commander/nav2_simple_commander/costmap_2d.py]: I split the getCost into getCostIdx and getCostXY

File: nav2_simple_commander/nav2_simple_commander/costmap_2d.py (hunk 2 above)
Comment (JacksonK9) [reply nav2_simple_commander/nav2_simple_commander/costmap_2d.py:4]: It is not for the company, So I think don't need to change

**Comment by JacksonK9** [comment]
I change the all requests, Please Review again.
After Update the https://navigation.ros.org/commander_api/index.html, I will notify again.

**Comment by JacksonK9** [comment]
I finished to update documentations

related Pull Request for documentation

* ros-planning/navigation.ros.org#344

**Comment by JacksonK9** [comment]
Thank you for your comment.

Lint code with flake8 😄

Remove Default value of getCostXY

update docstring like navigation.ros.org documentation.

Thank you for your review always 👍

File: nav2_simple_commander/nav2_simple_commander/costmap_2d.py [hunk 3]
```diff
@@ -33,8 +34,7 @@ def __init__(self, occupancy_map):
         self.global_frame_id = occupancy_map.header.frame_id
         self.costmap_timestamp = occupancy_map.header.stamp
         # Extract costmap
-        self.costmap = np.array(occupancy_map.data, dtype=np.int8).reshape(
-            self.size_y, self.size_x)
+        self.costmap = np.array(occupancy_map.data, dtype=np.int8)
```
Comment (SteveMacenski) [inline nav2_simple_commander/nav2_simple_commander/costmap_2d.py:37]: should be unsigned?

**Comment by SteveMacenski** [comment]
pep8 is still reporting linting errors https://app.circleci.com/pipelines/github/ros-planning/navigation2/7879/workflows/4b47923b-9136-47d8-81f1-07df10f1e609/jobs/27120

File: nav2_simple_commander/nav2_simple_commander/costmap_2d.py (hunk 3 above)
Comment (JacksonK9) [reply nav2_simple_commander/nav2_simple_commander/costmap_2d.py:37]: I refer to the cosmap_2d cpp code. 
[costmap_2d.hpp](https://github.com/ros-planning/navigation2/blob/main/nav2_costmap_2d/include/nav2_costmap_2d/costmap_2d.hpp)
The return value of costmap is unsigned char.
I think the python should also use the uint type.
What do you think?

Comment (SteveMacenski) [reply nav2_simple_commander/nav2_simple_commander/costmap_2d.py:37]: Agreed

Comment (JacksonK9) [reply nav2_simple_commander/nav2_simple_commander/costmap_2d.py:37]: Okay I will change and notify again 😄

**Review by SteveMacenski — APPROVED** [review]

**Pull request merged**